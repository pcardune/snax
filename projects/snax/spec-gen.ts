type BooleanLiteralFields = { value: boolean };

export type BooleanLiteral = {
  name: 'BooleanLiteral';
  fields: BooleanLiteralFields;
};

export function isBooleanLiteral(node: ASTNode): node is BooleanLiteral {
  return node.name === 'BooleanLiteral';
}

export function makeBooleanLiteral(value: boolean): BooleanLiteral {
  return {
    name: 'BooleanLiteral',
    fields: {
      value,
    },
  };
}

export function makeBooleanLiteralWith(fields: {
  value: boolean;
}): BooleanLiteral {
  return {
    name: 'BooleanLiteral',
    fields,
  };
}

type NumberLiteralFields = {
  value: number;
  numberType: 'int' | 'float';
  explicitType?: string;
};

export type NumberLiteral = {
  name: 'NumberLiteral';
  fields: NumberLiteralFields;
};

export function isNumberLiteral(node: ASTNode): node is NumberLiteral {
  return node.name === 'NumberLiteral';
}

export function makeNumberLiteral(
  value: number,
  numberType: 'int' | 'float',
  explicitType: string | undefined
): NumberLiteral {
  return {
    name: 'NumberLiteral',
    fields: {
      value,
      numberType,
      explicitType,
    },
  };
}

export function makeNumberLiteralWith(fields: {
  value: number;
  numberType: 'int' | 'float';
  explicitType?: string;
}): NumberLiteral {
  return {
    name: 'NumberLiteral',
    fields,
  };
}

type StringLiteralFields = { value: string };

export type StringLiteral = {
  name: 'StringLiteral';
  fields: StringLiteralFields;
};

export function isStringLiteral(node: ASTNode): node is StringLiteral {
  return node.name === 'StringLiteral';
}

export function makeStringLiteral(value: string): StringLiteral {
  return {
    name: 'StringLiteral',
    fields: {
      value,
    },
  };
}

export function makeStringLiteralWith(fields: {
  value: string;
}): StringLiteral {
  return {
    name: 'StringLiteral',
    fields,
  };
}

type SymbolRefFields = { symbol: string };

export type SymbolRef = {
  name: 'SymbolRef';
  fields: SymbolRefFields;
};

export function isSymbolRef(node: ASTNode): node is SymbolRef {
  return node.name === 'SymbolRef';
}

export function makeSymbolRef(symbol: string): SymbolRef {
  return {
    name: 'SymbolRef',
    fields: {
      symbol,
    },
  };
}

export function makeSymbolRefWith(fields: { symbol: string }): SymbolRef {
  return {
    name: 'SymbolRef',
    fields,
  };
}

type TypeRefFields = { symbol: string };

export type TypeRef = {
  name: 'TypeRef';
  fields: TypeRefFields;
};

export function isTypeRef(node: ASTNode): node is TypeRef {
  return node.name === 'TypeRef';
}

export function makeTypeRef(symbol: string): TypeRef {
  return {
    name: 'TypeRef',
    fields: {
      symbol,
    },
  };
}

export function makeTypeRefWith(fields: { symbol: string }): TypeRef {
  return {
    name: 'TypeRef',
    fields,
  };
}

type PointerTypeExprFields = { pointerToExpr: TypeExpr };

export type PointerTypeExpr = {
  name: 'PointerTypeExpr';
  fields: PointerTypeExprFields;
};

export function isPointerTypeExpr(node: ASTNode): node is PointerTypeExpr {
  return node.name === 'PointerTypeExpr';
}

export function makePointerTypeExpr(pointerToExpr: TypeExpr): PointerTypeExpr {
  return {
    name: 'PointerTypeExpr',
    fields: {
      pointerToExpr,
    },
  };
}

export function makePointerTypeExprWith(fields: {
  pointerToExpr: TypeExpr;
}): PointerTypeExpr {
  return {
    name: 'PointerTypeExpr',
    fields,
  };
}

type GlobalDeclFields = {
  symbol: string;
  typeExpr?: TypeExpr;
  expr: Expression;
};

export type GlobalDecl = {
  name: 'GlobalDecl';
  fields: GlobalDeclFields;
};

export function isGlobalDecl(node: ASTNode): node is GlobalDecl {
  return node.name === 'GlobalDecl';
}

export function makeGlobalDecl(
  symbol: string,
  typeExpr: TypeExpr | undefined,
  expr: Expression
): GlobalDecl {
  return {
    name: 'GlobalDecl',
    fields: {
      symbol,
      typeExpr,
      expr,
    },
  };
}

export function makeGlobalDeclWith(fields: {
  symbol: string;
  typeExpr?: TypeExpr;
  expr: Expression;
}): GlobalDecl {
  return {
    name: 'GlobalDecl',
    fields,
  };
}

type LetStatementFields = {
  symbol: string;
  typeExpr?: TypeExpr;
  expr: Expression;
};

export type LetStatement = {
  name: 'LetStatement';
  fields: LetStatementFields;
};

export function isLetStatement(node: ASTNode): node is LetStatement {
  return node.name === 'LetStatement';
}

export function makeLetStatement(
  symbol: string,
  typeExpr: TypeExpr | undefined,
  expr: Expression
): LetStatement {
  return {
    name: 'LetStatement',
    fields: {
      symbol,
      typeExpr,
      expr,
    },
  };
}

export function makeLetStatementWith(fields: {
  symbol: string;
  typeExpr?: TypeExpr;
  expr: Expression;
}): LetStatement {
  return {
    name: 'LetStatement',
    fields,
  };
}

type IfStatementFields = {
  condExpr: Expression;
  thenBlock: Block;
  elseBlock: Block;
};

export type IfStatement = {
  name: 'IfStatement';
  fields: IfStatementFields;
};

export function isIfStatement(node: ASTNode): node is IfStatement {
  return node.name === 'IfStatement';
}

export function makeIfStatement(
  condExpr: Expression,
  thenBlock: Block,
  elseBlock: Block
): IfStatement {
  return {
    name: 'IfStatement',
    fields: {
      condExpr,
      thenBlock,
      elseBlock,
    },
  };
}

export function makeIfStatementWith(fields: {
  condExpr: Expression;
  thenBlock: Block;
  elseBlock: Block;
}): IfStatement {
  return {
    name: 'IfStatement',
    fields,
  };
}

type WhileStatementFields = { condExpr: Expression; thenBlock: Block };

export type WhileStatement = {
  name: 'WhileStatement';
  fields: WhileStatementFields;
};

export function isWhileStatement(node: ASTNode): node is WhileStatement {
  return node.name === 'WhileStatement';
}

export function makeWhileStatement(
  condExpr: Expression,
  thenBlock: Block
): WhileStatement {
  return {
    name: 'WhileStatement',
    fields: {
      condExpr,
      thenBlock,
    },
  };
}

export function makeWhileStatementWith(fields: {
  condExpr: Expression;
  thenBlock: Block;
}): WhileStatement {
  return {
    name: 'WhileStatement',
    fields,
  };
}

type BlockFields = { statements: Statement[] };

export type Block = {
  name: 'Block';
  fields: BlockFields;
};

export function isBlock(node: ASTNode): node is Block {
  return node.name === 'Block';
}

export function makeBlock(statements: Statement[]): Block {
  return {
    name: 'Block',
    fields: {
      statements,
    },
  };
}

export function makeBlockWith(fields: { statements: Statement[] }): Block {
  return {
    name: 'Block',
    fields,
  };
}

type BinaryExprFields = { op: string; left: Expression; right: Expression };

export type BinaryExpr = {
  name: 'BinaryExpr';
  fields: BinaryExprFields;
};

export function isBinaryExpr(node: ASTNode): node is BinaryExpr {
  return node.name === 'BinaryExpr';
}

export function makeBinaryExpr(
  op: string,
  left: Expression,
  right: Expression
): BinaryExpr {
  return {
    name: 'BinaryExpr',
    fields: {
      op,
      left,
      right,
    },
  };
}

export function makeBinaryExprWith(fields: {
  op: string;
  left: Expression;
  right: Expression;
}): BinaryExpr {
  return {
    name: 'BinaryExpr',
    fields,
  };
}

type CallExprFields = { left: Expression; right: ArgList };

export type CallExpr = {
  name: 'CallExpr';
  fields: CallExprFields;
};

export function isCallExpr(node: ASTNode): node is CallExpr {
  return node.name === 'CallExpr';
}

export function makeCallExpr(left: Expression, right: ArgList): CallExpr {
  return {
    name: 'CallExpr',
    fields: {
      left,
      right,
    },
  };
}

export function makeCallExprWith(fields: {
  left: Expression;
  right: ArgList;
}): CallExpr {
  return {
    name: 'CallExpr',
    fields,
  };
}

type CastExprFields = { left: Expression; right: TypeExpr };

export type CastExpr = {
  name: 'CastExpr';
  fields: CastExprFields;
};

export function isCastExpr(node: ASTNode): node is CastExpr {
  return node.name === 'CastExpr';
}

export function makeCastExpr(left: Expression, right: TypeExpr): CastExpr {
  return {
    name: 'CastExpr',
    fields: {
      left,
      right,
    },
  };
}

export function makeCastExprWith(fields: {
  left: Expression;
  right: TypeExpr;
}): CastExpr {
  return {
    name: 'CastExpr',
    fields,
  };
}

type UnaryExprFields = { op: string; expr: Expression };

export type UnaryExpr = {
  name: 'UnaryExpr';
  fields: UnaryExprFields;
};

export function isUnaryExpr(node: ASTNode): node is UnaryExpr {
  return node.name === 'UnaryExpr';
}

export function makeUnaryExpr(op: string, expr: Expression): UnaryExpr {
  return {
    name: 'UnaryExpr',
    fields: {
      op,
      expr,
    },
  };
}

export function makeUnaryExprWith(fields: {
  op: string;
  expr: Expression;
}): UnaryExpr {
  return {
    name: 'UnaryExpr',
    fields,
  };
}

type ArrayLiteralFields = { elements: Expression[] };

export type ArrayLiteral = {
  name: 'ArrayLiteral';
  fields: ArrayLiteralFields;
};

export function isArrayLiteral(node: ASTNode): node is ArrayLiteral {
  return node.name === 'ArrayLiteral';
}

export function makeArrayLiteral(elements: Expression[]): ArrayLiteral {
  return {
    name: 'ArrayLiteral',
    fields: {
      elements,
    },
  };
}

export function makeArrayLiteralWith(fields: {
  elements: Expression[];
}): ArrayLiteral {
  return {
    name: 'ArrayLiteral',
    fields,
  };
}

type ParameterListFields = { parameters: Parameter[] };

export type ParameterList = {
  name: 'ParameterList';
  fields: ParameterListFields;
};

export function isParameterList(node: ASTNode): node is ParameterList {
  return node.name === 'ParameterList';
}

export function makeParameterList(parameters: Parameter[]): ParameterList {
  return {
    name: 'ParameterList',
    fields: {
      parameters,
    },
  };
}

export function makeParameterListWith(fields: {
  parameters: Parameter[];
}): ParameterList {
  return {
    name: 'ParameterList',
    fields,
  };
}

type ParameterFields = { symbol: string; typeExpr: TypeExpr };

export type Parameter = {
  name: 'Parameter';
  fields: ParameterFields;
};

export function isParameter(node: ASTNode): node is Parameter {
  return node.name === 'Parameter';
}

export function makeParameter(symbol: string, typeExpr: TypeExpr): Parameter {
  return {
    name: 'Parameter',
    fields: {
      symbol,
      typeExpr,
    },
  };
}

export function makeParameterWith(fields: {
  symbol: string;
  typeExpr: TypeExpr;
}): Parameter {
  return {
    name: 'Parameter',
    fields,
  };
}

type FuncDeclFields = {
  symbol: string;
  parameters: ParameterList;
  returnType?: TypeExpr;
  body: Block;
};

export type FuncDecl = {
  name: 'FuncDecl';
  fields: FuncDeclFields;
};

export function isFuncDecl(node: ASTNode): node is FuncDecl {
  return node.name === 'FuncDecl';
}

export function makeFuncDecl(
  symbol: string,
  parameters: ParameterList,
  returnType: TypeExpr | undefined,
  body: Block
): FuncDecl {
  return {
    name: 'FuncDecl',
    fields: {
      symbol,
      parameters,
      returnType,
      body,
    },
  };
}

export function makeFuncDeclWith(fields: {
  symbol: string;
  parameters: ParameterList;
  returnType?: TypeExpr;
  body: Block;
}): FuncDecl {
  return {
    name: 'FuncDecl',
    fields,
  };
}

type ReturnStatementFields = { expr: Expression };

export type ReturnStatement = {
  name: 'ReturnStatement';
  fields: ReturnStatementFields;
};

export function isReturnStatement(node: ASTNode): node is ReturnStatement {
  return node.name === 'ReturnStatement';
}

export function makeReturnStatement(expr: Expression): ReturnStatement {
  return {
    name: 'ReturnStatement',
    fields: {
      expr,
    },
  };
}

export function makeReturnStatementWith(fields: {
  expr: Expression;
}): ReturnStatement {
  return {
    name: 'ReturnStatement',
    fields,
  };
}

type ExprStatementFields = { expr: Expression };

export type ExprStatement = {
  name: 'ExprStatement';
  fields: ExprStatementFields;
};

export function isExprStatement(node: ASTNode): node is ExprStatement {
  return node.name === 'ExprStatement';
}

export function makeExprStatement(expr: Expression): ExprStatement {
  return {
    name: 'ExprStatement',
    fields: {
      expr,
    },
  };
}

export function makeExprStatementWith(fields: {
  expr: Expression;
}): ExprStatement {
  return {
    name: 'ExprStatement',
    fields,
  };
}

type ArgListFields = { args: Expression[] };

export type ArgList = {
  name: 'ArgList';
  fields: ArgListFields;
};

export function isArgList(node: ASTNode): node is ArgList {
  return node.name === 'ArgList';
}

export function makeArgList(args: Expression[]): ArgList {
  return {
    name: 'ArgList',
    fields: {
      args,
    },
  };
}

export function makeArgListWith(fields: { args: Expression[] }): ArgList {
  return {
    name: 'ArgList',
    fields,
  };
}

type FileFields = {
  funcs: FuncDecl[];
  globals: GlobalDecl[];
  decls: ExternDecl[];
};

export type File = {
  name: 'File';
  fields: FileFields;
};

export function isFile(node: ASTNode): node is File {
  return node.name === 'File';
}

export function makeFile(
  funcs: FuncDecl[],
  globals: GlobalDecl[],
  decls: ExternDecl[]
): File {
  return {
    name: 'File',
    fields: {
      funcs,
      globals,
      decls,
    },
  };
}

export function makeFileWith(fields: {
  funcs: FuncDecl[];
  globals: GlobalDecl[];
  decls: ExternDecl[];
}): File {
  return {
    name: 'File',
    fields,
  };
}

type ExternDeclFields = { libName: string; funcs: FuncDecl[] };

export type ExternDecl = {
  name: 'ExternDecl';
  fields: ExternDeclFields;
};

export function isExternDecl(node: ASTNode): node is ExternDecl {
  return node.name === 'ExternDecl';
}

export function makeExternDecl(libName: string, funcs: FuncDecl[]): ExternDecl {
  return {
    name: 'ExternDecl',
    fields: {
      libName,
      funcs,
    },
  };
}

export function makeExternDeclWith(fields: {
  libName: string;
  funcs: FuncDecl[];
}): ExternDecl {
  return {
    name: 'ExternDecl',
    fields,
  };
}

export type TypeExpr = PointerTypeExpr | TypeRef;
export function isTypeExpr(node: ASTNode): node is TypeExpr {
  return isPointerTypeExpr(node) || isTypeRef(node);
}

export type LiteralExpr =
  | NumberLiteral
  | StringLiteral
  | ArrayLiteral
  | BooleanLiteral
  | SymbolRef;
export function isLiteralExpr(node: ASTNode): node is LiteralExpr {
  return (
    isNumberLiteral(node) ||
    isStringLiteral(node) ||
    isArrayLiteral(node) ||
    isBooleanLiteral(node) ||
    isSymbolRef(node)
  );
}

export type Expression =
  | BinaryExpr
  | UnaryExpr
  | LiteralExpr
  | CallExpr
  | CastExpr
  | ArgList;
export function isExpression(node: ASTNode): node is Expression {
  return (
    isBinaryExpr(node) ||
    isUnaryExpr(node) ||
    isLiteralExpr(node) ||
    isCallExpr(node) ||
    isCastExpr(node) ||
    isArgList(node)
  );
}

export type Statement =
  | ReturnStatement
  | WhileStatement
  | IfStatement
  | LetStatement
  | ExprStatement
  | Block;
export function isStatement(node: ASTNode): node is Statement {
  return (
    isReturnStatement(node) ||
    isWhileStatement(node) ||
    isIfStatement(node) ||
    isLetStatement(node) ||
    isExprStatement(node) ||
    isBlock(node)
  );
}
export type ASTNode =
  | BooleanLiteral
  | NumberLiteral
  | StringLiteral
  | SymbolRef
  | TypeRef
  | PointerTypeExpr
  | GlobalDecl
  | LetStatement
  | IfStatement
  | WhileStatement
  | Block
  | BinaryExpr
  | CallExpr
  | CastExpr
  | UnaryExpr
  | ArrayLiteral
  | ParameterList
  | Parameter
  | FuncDecl
  | ReturnStatement
  | ExprStatement
  | ArgList
  | File
  | ExternDecl
  | TypeExpr
  | LiteralExpr
  | Expression
  | Statement;
export type ASTNodeName =
  | 'BooleanLiteral'
  | 'NumberLiteral'
  | 'StringLiteral'
  | 'SymbolRef'
  | 'TypeRef'
  | 'PointerTypeExpr'
  | 'GlobalDecl'
  | 'LetStatement'
  | 'IfStatement'
  | 'WhileStatement'
  | 'Block'
  | 'BinaryExpr'
  | 'CallExpr'
  | 'CastExpr'
  | 'UnaryExpr'
  | 'ArrayLiteral'
  | 'ParameterList'
  | 'Parameter'
  | 'FuncDecl'
  | 'ReturnStatement'
  | 'ExprStatement'
  | 'ArgList'
  | 'File'
  | 'ExternDecl';
