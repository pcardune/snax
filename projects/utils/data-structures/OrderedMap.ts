import { iter } from '../iter';

export class OrderedMap<K, V> {
  private keyMap: Map<K, V>;
  private keyList: K[];

  constructor(pairs: Iterable<[K, V]> = []) {
    this.keyMap = new Map();
    this.keyList = [];
    for (const [key, value] of pairs) {
      this.push(key, value);
    }
  }

  get length() {
    return this.keyList.length;
  }

  set(key: K, value: V) {
    if (this.keyMap.get(key) === undefined) {
      this.push(key, value);
    } else {
      this.keyMap.set(key, value);
    }
  }

  has(key: K) {
    return this.keyMap.has(key);
  }

  delete(key: K) {
    this.keyMap.delete(key);
    this.keyList.splice(this.keyList.indexOf(key), 1);
  }

  get(key: K) {
    return this.keyMap.get(key);
  }

  getAt(index: number): V | undefined {
    return this.keyMap.get(this.keyList[index]);
  }

  indexOf(key: K) {
    this.keyList.indexOf(key);
  }

  findIndex(callbackFn: (k: K, v: V) => boolean) {
    for (const [i, k, v] of this.entries()) {
      if (callbackFn(k, v)) {
        return i;
      }
    }
  }

  findKey(callbackFn: (v: V) => boolean) {
    for (const [i, k, v] of this.entries()) {
      if (callbackFn(v)) {
        return k;
      }
    }
  }

  push(key: K, value: V) {
    if (this.keyMap.get(key) !== undefined) {
      throw new Error(`key ${key} already in map`);
    }
    this.keyMap.set(key, value);
    this.keyList.push(key);
  }

  keys() {
    return iter(this.keyList);
  }

  *values() {
    for (const [i, key] of this.keyList.entries()) {
      yield this.get(key) as V;
    }
  }

  entries() {
    let i = 0;
    return this.keys().map(
      (key) => [i++, key, this.get(key) as V] as [number, K, V]
    );
  }

  map<W>(f: (value: V, index: number, key: K) => W): OrderedMap<K, W> {
    const map: OrderedMap<K, W> = new OrderedMap();
    for (const [i, k, v] of this.entries()) {
      map.push(k, f(v, i, k));
    }
    return map;
  }
}
