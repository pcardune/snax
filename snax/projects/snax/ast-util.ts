import {
  ASTNode,
  Block,
  makeBlock,
  makeFuncDecl,
  makeFuncDeclWith,
  makeNumberLiteral,
  makeParameterList,
  Parameter,
  Statement,
} from './spec-gen.js';

export function makeNum(
  value: number,
  interpretation: 'int' | 'float' = 'int'
) {
  return makeNumberLiteral(value, interpretation, undefined);
}

export function makeFunc(
  name: string,
  params: Parameter[] = [],
  body: Statement[] | Block = []
) {
  return makeFuncDeclWith({
    symbol: name,
    isPublic: false,
    parameters: makeParameterList(params ?? []),
    returnType: undefined,
    body: body instanceof Array ? makeBlock(body) : body,
  });
}

export function getPropNameOrThrow(node: ASTNode): string {
  switch (node.name) {
    case 'SymbolRef':
      return node.fields.symbol;
    case 'NumberLiteral':
      return String(node.fields.value);
    default:
      throw new Error(`${node.name} is not a valid property node`);
  }
}
