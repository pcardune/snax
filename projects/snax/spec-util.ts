import { OrderedMap } from '../utils/data-structures/OrderedMap';
import { ASTNode } from './spec-gen';
import { SymbolTable } from './symbol-resolution';

function isASTNode(node: any): node is ASTNode {
  return typeof node === 'object' && node !== null && node.name && node.fields;
}

export function children(node: ASTNode): ASTNode[] {
  let children: ASTNode[] = [];
  if (isASTNode(node)) {
    for (let field of Object.values(node.fields)) {
      if (field instanceof Array) {
        children.push(...field);
      } else if (isASTNode(field)) {
        children.push(field);
      }
    }
  }
  return children;
}

export function* preorderIter(node: ASTNode): Generator<ASTNode> {
  yield node;
  for (const child of children(node)) {
    yield* preorderIter(child);
  }
}

export function* depthFirstIter(node: ASTNode): Generator<ASTNode> {
  for (const child of children(node)) {
    yield* depthFirstIter(child);
  }
  yield node;
}

export function debugStr(node: ASTNode) {
  let s = `<${node.name} `;
  for (let [fieldName, value] of Object.entries(node.fields)) {
    if (isASTNode(value) || value instanceof Array) {
      continue;
    }
    s += `${fieldName}=${JSON.stringify(value)} `;
  }
  s += '/>';
  return s;
}

type Elem = { tag: string; props: Record<string, any>; children: Elem[] };
function elem(tag: string, props: Record<string, any>, children: Elem[]): Elem {
  return {
    tag,
    props,
    children,
  };
}

function elemToString(elem: Elem, indent = '') {
  let s = `${indent}<${elem.tag}`;

  let props = Object.entries(elem.props)
    .map(([propName, propValue]) => `${propName}=${JSON.stringify(propValue)}`)
    .join(' ');
  if (props.length > 0) {
    s += ' ' + props;
  }
  if (elem.children.length == 0) {
    return s + '/>';
  }

  s += '>\n';

  for (const child of elem.children) {
    s += elemToString(child, indent + '  ') + '\n';
  }

  s += `${indent}</${elem.tag}>`;
  return s;
}

function nodeDataToElem(
  node: ASTNode,
  symbolTables: OrderedMap<ASTNode, SymbolTable>
) {
  let props: Record<string, any> = {};
  let childElems: Elem[] = [];

  let symbolTable = symbolTables.get(node);
  if (symbolTable) {
    childElems.push(
      elem(
        'SymbolTable',
        { id: symbolTable.id, parent: symbolTable.parent?.id },
        symbolTable.table
          .entries()
          .map(([i, symbol, record]) => {
            let props = {};
            return elem(symbol, props, []);
          })
          .toArray()
      )
    );
  }

  for (let [fieldName, value] of Object.entries(node.fields)) {
    if (isASTNode(value)) {
      childElems.push(nodeDataToElem(value, symbolTables));
    } else if (value instanceof Array) {
      childElems.push(
        elem(
          fieldName,
          {},
          value.map((v) => nodeDataToElem(v, symbolTables))
        )
      );
    } else {
      props[fieldName] = value;
    }
  }
  return elem(node.name, props, childElems);
}

export function dumpSymbolTables(
  node: ASTNode,
  symbolTables: OrderedMap<ASTNode, SymbolTable>
) {
  return elemToString(nodeDataToElem(node, symbolTables));
}
