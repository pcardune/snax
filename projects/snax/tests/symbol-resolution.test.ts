import * as AST from '../spec-gen';
import { makeFunc, makeNum } from './ast-util';
import { dumpSymbolTables } from '../spec-util';
import {
  resolveSymbols,
  SymbolRecord,
  SymbolTable,
} from '../symbol-resolution';
import { OrderedMap } from '../../utils/data-structures/OrderedMap';

let file: AST.File;
let funcDecl: AST.FuncDecl;
let globalDecl: AST.GlobalDecl;
let outerBlock: AST.Block;
let innerBlock: AST.Block;
let innerXRef: AST.SymbolRef;
let innerYRef: AST.SymbolRef;
let outerXRef: AST.SymbolRef;
let tables: OrderedMap<AST.ASTNode, SymbolTable>;
let refMap: OrderedMap<AST.ASTNode, SymbolRecord>;
beforeEach(() => {
  innerXRef = AST.makeSymbolRef('x');
  innerYRef = AST.makeSymbolRef('y');
  innerBlock = AST.makeBlock([
    AST.makeLetStatement('x', null, makeNum(3)),
    AST.makeExprStatement(innerXRef),
    AST.makeExprStatement(innerYRef),
  ]);
  outerXRef = AST.makeSymbolRef('x');
  outerBlock = AST.makeBlock([
    AST.makeLetStatement('x', null, makeNum(1)),
    AST.makeLetStatement('y', null, makeNum(2)),
    innerBlock,
    AST.makeExprStatement(outerXRef),
  ]);
  globalDecl = AST.makeGlobalDecl('g', null, makeNum(10));
  funcDecl = makeFunc('main', [], outerBlock);
  file = AST.makeFile([funcDecl], [globalDecl]);
  const resolution = resolveSymbols(file);
  tables = resolution.tables;
  refMap = resolution.refMap;
});

describe('resolveSymbols', () => {
  it('constructs the correct symbol table', () => {
    let symbolTable = tables.get(outerBlock);
    if (!symbolTable) {
      fail('resolveSymbols should attach a symbolTable to the block');
    }
    expect(symbolTable.get('x')?.declNode).toBe(
      outerBlock.fields.statements[0]
    );
    expect(symbolTable.get('y')?.declNode).toBe(
      outerBlock.fields.statements[1]
    );
    expect(refMap.get(outerXRef)).toBe(symbolTable.get('x'));
  });
  it('symbols are resolved to the nearest scope they are found in', () => {
    expect(refMap.get(innerXRef)).not.toBe(tables.get(outerBlock)?.get('x'));
    expect(refMap.get(innerXRef)).toBe(tables.get(innerBlock)?.get('x'));
  });
  it('symbols are resolved to an outer scope when not in the current scope', () => {
    expect(refMap.get(innerYRef)).toBe(tables.get(outerBlock)?.get('y'));
  });
  it('globals appear in the files symbol table', () => {
    expect(tables.get(file)?.has('g')).toBe(true);
  });
  it('spits out the right debug info', () => {
    expect(dumpSymbolTables(file, tables)).toMatchInlineSnapshot(`
      "<File>
        <SymbolTable id=17 parent=16>
          <g/>
          <main/>
        </SymbolTable>
        <funcs>
          <FuncDecl symbol=\\"main\\" returnType=null>
            <ParameterList>
              <parameters/>
            </ParameterList>
            <Block>
              <SymbolTable id=18 parent=17>
                <x/>
                <y/>
              </SymbolTable>
              <statements>
                <LetStatement symbol=\\"x\\" typeExpr=null>
                  <NumberLiteral value=1 numberType=\\"int\\" explicitType=null/>
                </LetStatement>
                <LetStatement symbol=\\"y\\" typeExpr=null>
                  <NumberLiteral value=2 numberType=\\"int\\" explicitType=null/>
                </LetStatement>
                <Block>
                  <SymbolTable id=19 parent=18>
                    <x/>
                  </SymbolTable>
                  <statements>
                    <LetStatement symbol=\\"x\\" typeExpr=null>
                      <NumberLiteral value=3 numberType=\\"int\\" explicitType=null/>
                    </LetStatement>
                    <ExprStatement>
                      <SymbolRef symbol=\\"x\\"/>
                    </ExprStatement>
                    <ExprStatement>
                      <SymbolRef symbol=\\"y\\"/>
                    </ExprStatement>
                  </statements>
                </Block>
                <ExprStatement>
                  <SymbolRef symbol=\\"x\\"/>
                </ExprStatement>
              </statements>
            </Block>
          </FuncDecl>
        </funcs>
        <globals>
          <GlobalDecl symbol=\\"g\\" typeExpr=null>
            <NumberLiteral value=10 numberType=\\"int\\" explicitType=null/>
          </GlobalDecl>
        </globals>
      </File>"
    `);
  });
});
