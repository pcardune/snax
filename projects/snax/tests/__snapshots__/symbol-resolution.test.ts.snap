// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveSymbols spits out the right debug info 1`] = `
"<File>
  <SymbolTable id=21 parent=20>
    <externalFunc/>
    <g/>
    <main/>
  </SymbolTable>
  <funcs>
    <FuncDecl symbol=\\"main\\" returnType=undefined>
      <ParameterList>
        <parameters/>
      </ParameterList>
      <Block>
        <SymbolTable id=22 parent=21>
          <x/>
          <y/>
        </SymbolTable>
        <statements>
          <LetStatement symbol=\\"x\\" typeExpr=undefined>
            <NumberLiteral value=1 numberType=\\"int\\" explicitType=undefined/>
          </LetStatement>
          <LetStatement symbol=\\"y\\" typeExpr=undefined>
            <NumberLiteral value=2 numberType=\\"int\\" explicitType=undefined/>
          </LetStatement>
          <Block>
            <SymbolTable id=23 parent=22>
              <x/>
            </SymbolTable>
            <statements>
              <LetStatement symbol=\\"x\\" typeExpr=undefined>
                <NumberLiteral value=3 numberType=\\"int\\" explicitType=undefined/>
              </LetStatement>
              <ExprStatement>
                <SymbolRef symbol=\\"x\\"/>
              </ExprStatement>
              <ExprStatement>
                <SymbolRef symbol=\\"y\\"/>
              </ExprStatement>
            </statements>
          </Block>
          <ExprStatement>
            <SymbolRef symbol=\\"x\\"/>
          </ExprStatement>
        </statements>
      </Block>
    </FuncDecl>
  </funcs>
  <globals>
    <GlobalDecl symbol=\\"g\\" typeExpr=undefined>
      <NumberLiteral value=10 numberType=\\"int\\" explicitType=undefined/>
    </GlobalDecl>
  </globals>
  <decls>
    <ExternDecl libName=\\"someLib\\">
      <funcs>
        <FuncDecl symbol=\\"externalFunc\\" returnType=undefined>
          <ParameterList>
            <parameters/>
          </ParameterList>
          <Block>
            <statements/>
          </Block>
        </FuncDecl>
      </funcs>
    </ExternDecl>
  </decls>
</File>"
`;
